// angularfire2/es2015/firebase.app.module.js
import { NgModule } from '@angular/core';
import { FirebaseOptionsToken, FirebaseAppNameToken, FirebaseAppConfigToken } from './angularfire2';
import firebase from '@firebase/app';
export class FirebaseApp {
}
export function _firebaseAppFactory(options, name, appConfig) {
    const config = appConfig || {};
    if (name && config.name && config.name !== name) {
        console.warn('FirebaseAppNameToken and FirebaseAppConfigToken.name don\'t match, FirebaseAppNameToken takes precedence.');
    }
    config.name = name || config.name || '[DEFAULT]';
    const existingApp = firebase.apps.filter(app => app.name === config.name)[0];
    return (existingApp || firebase.initializeApp(options, config));
}
const FirebaseAppProvider = {
    provide: FirebaseApp,
    useFactory: _firebaseAppFactory,
    deps: [FirebaseOptionsToken, FirebaseAppNameToken, FirebaseAppConfigToken]
};
export class AngularFireModule {
    static initializeApp(options, appNameOrConfig) {
        const name = typeof appNameOrConfig === 'string' && appNameOrConfig || undefined;
        const config = typeof appNameOrConfig === 'object' && appNameOrConfig || undefined;
        return {
            ngModule: AngularFireModule,
            providers: [
                { provide: FirebaseOptionsToken, useValue: options },
                { provide: FirebaseAppNameToken, useValue: name },
                { provide: FirebaseAppConfigToken, useValue: config }
            ]
        };
    }
}
AngularFireModule.decorators = [
    { type: NgModule, args: [{
                providers: [FirebaseAppProvider],
            },] },
];
